esp-hal/src/timer/mod.rs

```diff
@@ -1,6 +1,231 @@
+    fn load_value(&self, value: MicrosDurationU64);
+
+    /// Enable auto reload of the loaded value.
+    fn enable_auto_reload(&self, auto_reload: bool);
+
+    /// Enable or disable the timer's interrupt.
+    fn enable_interrupt(&self, state: bool);
+
+    /// Clear the timer's interrupt.
+    fn clear_interrupt(&self);
+
+    /// Has the timer triggered?
+    fn is_interrupt_set(&self) -> bool;
+
+    /// FIXME: This is (hopefully?) temporary...
+    fn set_alarm_active(&self, state: bool);
+}
+
+/// A one-shot timer.
+pub struct OneShotTimer<T> {
+    inner: T,
+}
+
+impl<T> OneShotTimer<T>
+where
+    T: Timer,
+{
+    /// Construct a new instance of [`OneShotTimer`].
+    pub fn new(inner: T) -> Self {
+        Self { inner }
+    }
+
+    /// Pauses execution for *at least* `ms` milliseconds.
+    pub fn delay_millis(&self, ms: u32) {
+        self.delay((ms as u64).millis());
+    }
+
+    /// Pauses execution for *at least* `us` microseconds.
+    pub fn delay_micros(&self, us: u32) {
+        self.delay((us as u64).micros());
+    }
+
+    /// Pauses execution for *at least* `ns` nanoseconds.
+    pub fn delay_nanos(&self, ns: u32) {
+        self.delay((ns as u64 / 1000).micros())
+    }
+
+    fn delay(&self, us: MicrosDurationU64) {
+        if self.inner.is_running() {
+            self.inner.stop();
+        }
+
+        self.inner.clear_interrupt();
+        self.inner.reset();
+
+        // NOTE: Auto-reload MUST be enabled before loading the value in order
+        //       for `SYSTIMER` to work correctly!
+        self.inner.enable_auto_reload(false);
+        self.inner.load_value(us);
+        self.inner.enable_interrupt(true);
+        self.inner.start();
+
+        while !self.inner.is_interrupt_set() {
+            // Wait
+        }
+
+        self.inner.stop();
+        self.inner.clear_interrupt();
+    }
+}
+
+#[cfg(feature = "embedded-hal-02")]
+impl<T, UXX> embedded_hal_02::blocking::delay::DelayMs<UXX> for OneShotTimer<T>
+where
+    T: Timer,
+    UXX: Into<u32>,
+{
+    fn delay_ms(&mut self, ms: UXX) {
+        self.delay_millis(ms.into());
+    }
+}
+
+#[cfg(feature = "embedded-hal-02")]
+impl<T, UXX> embedded_hal_02::blocking::delay::DelayUs<UXX> for OneShotTimer<T>
+where
+    T: Timer,
+    UXX: Into<u32>,
+{
+    fn delay_us(&mut self, us: UXX) {
+        self.delay_micros(us.into());
+    }
+}
+
+#[cfg(feature = "embedded-hal")]
+impl<T> embedded_hal::delay::DelayNs for OneShotTimer<T>
+where
+    T: Timer,
+{
+    #[allow(clippy::useless_conversion)]
+    fn delay_ns(&mut self, ns: u32) {
+        self.delay_nanos(ns.into());
+    }
+}
+
+/// A periodic timer.
+pub struct PeriodicTimer<T> {
+    inner: T,
+}
+
+impl<T> PeriodicTimer<T>
+where
+    T: Timer,
+{
+    /// Construct a new instance of [`PeriodicTimer`].
+    pub fn new(inner: T) -> Self {
+        Self { inner }
+    }
+
+    /// Start a new count down.
+    pub fn start(&mut self, timeout: MicrosDurationU64) {
```

Comment:
should we mention that "too large" periods won't work? e.g. on C6 even 5000 millis is too much and exceeds the 26 bit period field

Also wondering if we should panic when exceeding the period? Not nice but depending on what the user code does might be hard to debug if the period isn't correct 🤔

Reply:
Maybe not panic, but we should return an error at the very least.

Reply:
sure ... what the API GUIDELINE suggests 😄

Reply:
As discussed elsewhere, this isn't as straight forward as it seems at first. Any issue with punting this off until the next PR? I think this is good to go otherwise now.

Reply:
Okay, let's punt this to 3/3!